aws/resource_aws_gamelift_queue.go

```diff
@@ -0,0 +1,200 @@
+		Names: aws.StringSlice([]string{d.Get("name").(string)}),
+		Limit: &limit,
+	})
+	if err != nil {
+		if isAWSErr(err, gamelift.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Gamelift Session Queues (%s) not found, removing from state", d.Get("name"))
+			return nil
+		}
+		return err
+	}
+	sessionQueues := out.GameSessionQueues
+
+	if len(sessionQueues) < 1 {
+		log.Printf("[WARN] Gamelift Session Queue (%s) not found, removing from state", d.Get("name"))
+		return nil
+	}
+	if len(sessionQueues) != 1 {
+		return fmt.Errorf("expected exactly 1 Gamelift Session Queues, found %d under %q",
+			len(sessionQueues), d.Get("name"))
+	}
+	sessionQueue := sessionQueues[0]
+
+	d.Set("destinations", sessionQueue.Destinations)
+	d.Set("name", sessionQueue.Name)
+	d.Set("player_latency_policies", sessionQueue.PlayerLatencyPolicies)
+	d.Set("timeout_in_seconds", sessionQueue.TimeoutInSeconds)
+
+	return nil
+}
+
+func resourceAwsGameliftQueueUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+
+	name := d.Get("name").(string)
+
+	log.Printf("[INFO] Updating Gamelift Session Queue: %s", name)
+
+	if d.HasChange("name") || d.HasChange("destinations") ||
+		d.HasChange("player_latency_policies") || d.HasChange("timeout_in_seconds") {
+
+		input := getFullInputUpdate(d)
+
+		_, err := conn.UpdateGameSessionQueue(&input)
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceAwsGameliftQueueRead(d, meta)
+}
+
+func resourceAwsGameliftQueueDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+	name := d.Get("name").(string)
+	log.Printf("[INFO] Deleting Gamelift Session Queue: %s", name)
+	_, err := conn.DeleteGameSessionQueue(&gamelift.DeleteGameSessionQueueInput{
+		Name: aws.String(name),
+	})
+	if err != nil {
+		return err
+	}
+
+	d.SetId("")
+	return nil
+
+}
+
+func getFullInputCreate(d *schema.ResourceData) gamelift.CreateGameSessionQueueInput {
+	return gamelift.CreateGameSessionQueueInput{
+		Name:                  aws.String(d.Get("name").(string)),
+		Destinations:          getDestinations(d.Get("destinations").([]interface{})),
+		PlayerLatencyPolicies: getPlayerLatencyPolicies(d.Get("player_latency_policies").([]interface{})),
+		TimeoutInSeconds:      aws.Int64(int64(d.Get("timeout_in_seconds").(int))),
+	}
+}
+
+func getFullInputUpdate(d *schema.ResourceData) gamelift.UpdateGameSessionQueueInput {
+	return gamelift.UpdateGameSessionQueueInput{
+		Name:                  aws.String(d.Get("name").(string)),
+		Destinations:          getDestinations(d.Get("destinations").([]interface{})),
+		PlayerLatencyPolicies: getPlayerLatencyPolicies(d.Get("player_latency_policies").([]interface{})),
+		TimeoutInSeconds:      aws.Int64(int64(d.Get("timeout_in_seconds").(int))),
+	}
+}
+
+func getDestinations(destinationsMap []interface{}) []*gamelift.GameSessionQueueDestination {
```

Comment:
Nit: Since this project has a very flat Go package structure we prefer to include the AWS service name in the function name. We also generally refer to these types of functions as "expand" functions, e.g. `expandGameliftGameSessionQueueDestinations`